docs: add reference snippets for deployment

edibotopic opened the pull request:
An important use-case for authd is **deploying at scale** when, for example, an organisation has multiple user machines that need to authenticate with a cloud provider.

This PR adds **reference pages with code snippets** for remote script execution using Landscape and provisioning with cloud-init.

UDENG-7785

adombeck commented on docs/reference/code/readme:
```diff
@@ -0,0 +1,106 @@
+# Testing documented authd deployment script
```
Adding a `.md` suffix to the filename would allow us to view the rendered markdown in GitHub. Without that extension, GitHub only shows it in plaintext: https://github.com/ubuntu/authd/blob/cca3f8ac974a2d3a9b5b7e505de63890fb410d12/docs/reference/code/readme

adombeck commented on docs/reference/landscape-script.md:
```diff
@@ -0,0 +1,82 @@
+To install the authd deb and the broker snap:
+
+:::{literalinclude} ./code/deploy_authd/tests/deploy_authd_landscape/task.yaml
+:language: text
```
```suggestion
:language: shell
```

adombeck commented on docs/reference/landscape-script.md:
```diff
@@ -0,0 +1,82 @@
+To configure authd and the broker:
+
+:::{literalinclude} ./code/deploy_authd/tests/deploy_authd_landscape/task.yaml
+:language: text
```
```suggestion
:language: shell
```

adombeck reviewed (commented):
It's really cool that you got this to work, and it's great for testing the command snippets on a local machine. 

However, I'm not sure if the spread + multipass approach is a suitable solution for CI, which is arguably the more important use case, to continuously test that the docs are up to date. In the CI, we don't need to fetch and launch a VM, because we're already running on a disposable system of our choice. Downloading a VM image each time in the CI is slow and resource-intensive. So I would prefer a solution which allows us to just execute the command blocks on the current system. Then for running the test on a local machine, we can spin up a VM or container and execute the test in there (which could just be done in a shell script, I don't think we need something like spread for that).

I did a quick search for a tool which extracts code blocks from markdown to execute them, and [there](https://gist.github.com/szkiba/d9b7b4601efb7dc68aa2c58548b8b271) [are](https://github.com/nschloe/pytest-codeblocks) [a](https://github.com/Widdershin/markdown-doctest) [few](https://percona.community/blog/2023/02/28/doc-testing/), but nothing that seems to be widely used or an obvious ideal candidate, so we might need to do some more research and try things out.

Anyway, I don't consider having tests a blocker to merging the cloud-init and landscape documentation, especially since they do indeed closely correspond to the documentation we already have.

edibotopic commented:
Thanks @adombeck -- I'll finalise next pulse :+1:

edibotopic commented:
@adombeck --- I'm marking this as ready to review.

I removed the automated testing and refactored into a standard doc, with tabs for the brokers like elsewhere.

I also added an equivalent reference page for cloud init.

adombeck commented on docs/reference/landscape-deploy.md:
````diff
@@ -0,0 +1,158 @@
+```shell
+add-apt-repository -y ppa:ubuntu-enterprise-desktop/authd
+apt update && apt upgrade -y
+apt-get install -y authd gnome-shell yaru-theme-gnome-shell
````
Why do we only install gnome-shell and yaru-theme-gnome-shell in the Entra ID tab and not in the Google tab?

If we want to install it or not depends on whether the system is Ubuntu Desktop or Ubuntu Server. Maybe you intended to name the tabs "Ubuntu Desktop" and "Ubuntu Server", same as in https://documentation.ubuntu.com/authd/stable-docs/howto/install-authd/#install-authd?

edibotopic replied:
It should be both. Will fix.

edibotopic replied:
From other comments in https://github.com/ubuntu/authd/pull/1093#discussion_r2398882467

Should the installation of gnome-shell and yaru-theme-gnome-shell be removed from the script?

adombeck replied:
yes, the `apt upgrade -y` should indeed be enough, and now we also identified a purpose for it :)

we don't need the `apt update` though.

edibotopic replied:
Only `upgrade` is included since https://github.com/ubuntu/authd/pull/1093/commits/cf728aa7ea4eedd495bad9bd56316902bef454c7

adombeck commented on docs/reference/landscape-deploy.md:
````diff
@@ -0,0 +1,158 @@
+:sync: google
+
+```shell
+systemctl restart authd
+snap restart authd-google
+systemctl restart ssh
````
```suggestion
systemctl restart authd ssh
snap restart authd-google
```

We can restart multiple services with a single command

adombeck commented on docs/reference/landscape-deploy.md:
```diff
@@ -0,0 +1,158 @@
+:::::
+
+When you have a complete script, add it to the Landscape dashboard to run as
+root user  before executing on the target machines.
```
```suggestion
root user before executing on the target machines.
```

adombeck commented on docs/reference/landscape-deploy.md:
```diff
@@ -0,0 +1,158 @@
+sed -i "s|<CLIENT_ID>|$CLIENT_ID|g; s|<ISSUER_ID>|$ISSUER_ID|g" /var/snap/authd-msentraid/current/broker.conf
+echo "ssh_allowed_suffixes = @test.onmicrosoft.com" >> /var/snap/authd-msentraid/current/broker.conf
+touch /etc/ssh/sshd_config.d/authd.conf
+echo "UsePAM yes" >> /etc/ssh/sshd_config.d/authd.conf
+echo "KbdInteractiveAuthentication yes" >> /etc/ssh/sshd_config.d/authd.conf
```
Similar to above:

```suggestion
sudo tee /etc/ssh/sshd_config.d/authd.conf > /dev/null <<EOF
UsePAM yes
KbdInteractiveAuthentication yes
EOF
```

adombeck commented on docs/reference/landscape-deploy.md:
````diff
@@ -0,0 +1,158 @@
+
+```shell
+sed -i "s|<CLIENT_ID>|$CLIENT_ID|g; s|<ISSUER_ID>|$ISSUER_ID|g" /var/snap/authd-msentraid/current/broker.conf
+echo "ssh_allowed_suffixes = @test.onmicrosoft.com" >> /var/snap/authd-msentraid/current/broker.conf
````
We could also use sed to insert the line below the existing `#ssh_allowed_suffixes = [...]` line:

```suggestion
sudo sed -i '/^#ssh_allowed_suffixes/ a ssh_allowed_suffixes = @test.onmicrosoft.com' /var/snap/authd-msentraid/current/broker.conf
```

edibotopic replied:
I think I have a brief note about running as root at the end, but it might be beneficial to mention it at the start for context.

We just don't want to devote too much space to documenting the Landscape behemoth.

edibotopic replied:
Kept note at end, rewriting slightly, and also added a note at the start: https://github.com/ubuntu/authd/pull/1093/commits/81e3e2238d97bdeaf98918429dcc3d6440017a8e

adombeck commented on docs/reference/landscape-deploy.md:
```diff
@@ -0,0 +1,158 @@
+add-apt-repository -y ppa:ubuntu-enterprise-desktop/authd
+apt update && apt upgrade -y
+apt-get install -y authd
+snap install authd-google
```
Ah okay, that changes things. Then indeed we don't need sudo and can use `echo >> <file>` to write to files.

edibotopic replied:
Tried to clarify this in https://github.com/ubuntu/authd/pull/1093/commits/81e3e2238d97bdeaf98918429dcc3d6440017a8e

adombeck commented on docs/reference/cloud-init-deploy.md:
```diff
@@ -0,0 +1,161 @@
+      KbdInteractiveAuthentication yes
+
+runcmd:
+  - sed -i 's|<CLIENT_ID>|{{ CLIENT_ID }}|g; s|<ISSUER_ID>|{{ ISSUER_ID }}|g' /var/snap/authd-msentraid/current/broker.conf
+  - echo 'ssh_allowed_suffixes = @test.onmicrosoft.com' >> /var/snap/authd-msentraid/current/broker.conf
+  - sed -i 's/^\(LOGIN_TIMEOUT\t\t\)[0-9]\+/\1360/' /etc/login.defs
+  - mkdir -p /etc/authd/brokers.d/
+  - cp /snap/authd-msentraid/current/conf/authd/msentraid.conf /etc/authd/brokers.d/
+  - snap restart authd-msentraid
+  - systemctl restart authd
+  - snap restart authd-msentraid
+  - systemctl restart ssh
```
Same changes as on the landscape page:

```suggestion
  - sed -i 's|<CLIENT_ID>|{{ CLIENT_ID }}|g; s|<ISSUER_ID>|{{ ISSUER_ID }}|g' /var/snap/authd-msentraid/current/broker.conf
  - echo 'ssh_allowed_suffixes = @example.onmicrosoft.com' >> /var/snap/authd-msentraid/current/broker.conf
  - sed -i 's/^\(LOGIN_TIMEOUT\t\t\)[0-9]\+/\1360/' /etc/login.defs
  - mkdir -p /etc/authd/brokers.d/
  - cp /snap/authd-msentraid/current/conf/authd/msentraid.conf /etc/authd/brokers.d/
  - snap restart authd-msentraid
  - systemctl restart authd ssh
```

3v1n0 commented on docs/reference/cloud-init-deploy.md:
```diff
@@ -0,0 +1,161 @@
+write_files:
+  - path: /etc/ssh/sshd_config.d/authd.conf
+    content: |
+      UsePAM yes
+      KbdInteractiveAuthentication yes
```
Maybe here we can suggest that they can use

```
Match *@domain.ext
    UsePAM yes
    KbdInteractiveAuthentication yes
```

To ensure that the configuration applies only to the specific user

adombeck replied:
Yes, wildcards are supported, see https://manpages.ubuntu.com/manpages/questing/en/man5/ssh_config.5.html#patterns

adombeck replied:
and I would prefer to stick to `example.com`, which is a reserved domain intended for documentation, which is ensured to not cause naming conflicts when its copied verbatim.

adombeck replied:
this works:

```
Match User *@example.com
    KbdInteractiveAuthentication yes
```

and if the global settings have the Debian default values

```
KbdInteractiveAuthentication no
UsePAM yes
```

Then it does still effectively prevent users who do not match the `*@example` pattern from logging in with a password.

adombeck replied:
Here I would just use

```
Match User *@example.com
    KbdInteractiveAuthentication yes
```

Users can then choose themselves if they want to deviate from the default config and use `PasswordAuthentication no` to disable password authentication for non-authd users. I added a section about that in the security overview.

adombeck replied:
Mmh maybe we should enforce `UsePAM yes` to ensure that login via authd works. So that snippet for `/etc/ssh/sshd_config.d/authd.conf` would become:

```
UsePAM yes
Match User *@example.com
    KbdInteractiveAuthentication yes
```

Do you have an opinion on that, @3v1n0?

adombeck replied:
> Mmh maybe we should enforce UsePAM yes to ensure that login via authd works. 

Eh, I now feel like this is too much bike shedding. Both with or without`UsePAM yes` in the config snippet is fine for me, just go with whatever you prefer Shane (from my side)

edibotopic commented on docs/reference/cloud-init-deploy.md:
```diff
@@ -0,0 +1,161 @@
+
+packages:
+  - authd
+  - gnome-shell # only needed for GDM login
+  - yaru-theme-gnome-shell # only needed for GDM login
```
I think `package_upgrade = true` at the top of the config does a `apt upgrade` rather than a `apt dist-upgrade`. 

Would that be sufficient, assuming it's already desktop environment?

edibotopic replied:
The most recent changes should make it so both happen after the PPA is added.

If it doesn't matter whether it's

```
runcmd:
  - apt-get install -y authd
  - apt-get upgrade -y
```

or


```
runcmd:
  - apt-get upgrade -y
  - apt-get install -y authd
```

I'll just leave it as the first.

adombeck replied:
but can't we keep

```
packages:
  - authd
```

instead of moving that to the run block?

adombeck replied:
> Then we should be done here... 👀

yes :)

edibotopic replied:
This OK?

https://github.com/ubuntu/authd/pull/1093/commits/4018a6b8558a82cc52241ada79093c65ddfd9b6a

edibotopic commented on docs/reference/landscape-deploy.md:
````diff
@@ -0,0 +1,158 @@
+
+```shell
+add-apt-repository -y ppa:ubuntu-enterprise-desktop/authd
+apt update && apt upgrade -y
````
Removed in https://github.com/ubuntu/authd/pull/1093/commits/a682ee0c7166c4734f9a072cd8a5ac67a80f5661

edibotopic commented on docs/reference/cloud-init-deploy.md:
````diff
@@ -0,0 +1,161 @@
+Define the necessary environmental variables:
+
+```text
+{% set ISSUER_ID = '<your_issuer_id>' %}
+{% set CLIENT_ID = '<your_client_id>' %}
````
Added an example header in https://github.com/ubuntu/authd/pull/1093/commits/7183dc34fddf9875aba72c893b5a544f91a73f46 that would enable Jinja for cloud-init. The header also includes the necessary `#cloud-config` and I added Jinja highlighting to the block.
